### DISCUSSION TITLE
How to properly type class abstract async methods and override them in child classes

### DISCUSSION THREAD

**Discussion by acartine** [microsoft/pyright#4741] [Q&A]
I have an abstract class with the method
```
    @abstractmethod
    async def reap(self, ts: int) -> None:
        pass
```

the child class implements it as follows

```
    async def reap(self, ts_ms: int) -> None:
        await self.__reap(ts_ms)
```

pyright reports the following

```
/home/acartine/code/nifty/src/nifty_worker/trend/toplist/async_toplist.py
  /home/acartine/code/nifty/src/nifty_worker/trend/toplist/async_toplist.py:154:15 - error: Method "reap" overrides class "AbstractTopList" in an incompatible manner
    Return type mismatch: base method returns type "Coroutine[Any, Any, None]", override returns type "Awaitable[None]"
      "Awaitable[None]" is incompatible with "Coroutine[Any, Any, None]" (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations 
```

what is the proper way to achieve what i'm trying to acheive here?  is it to define the child method without async, and have it return Coroutine[Any,Any,None] ?

**Comment by erictraut** [comment]
Could you provide a complete, minimal code sample that exhibits the problem? I tried to recreate the problem from your description, but I'm not receiving any error.

```python
# pyright: strict
from abc import ABC, abstractmethod

class Base(ABC):
    @abstractmethod
    async def reap(self, ts: int) -> None:
        pass

class Child(Base):
    async def reap(self, ts: int) -> None:
        pass
```

**Reply by acartine** [reply]
hey i went to go do so and could not.  it looks like i ran it against an an unsaved buffer.

however i was able to reproduce the error message with a function decorator above the `reap` implementation.  The function decorator had the return value

`Callable[..., Awaitable[_T]]`. 

so i assume that was the problem and that the typing of the decorator was incorrect.  When i changed it to 

`Callable[..., Coroutine[Any, Any, _T]]` 

pycheck was happy.

I was fiddling around with that and i must have had an unsaved buffer out of view when running my usual sanity checks.

**Comment by Kolpa** [comment]
It's been a while but i've just ran into what i think is the same issue running python `3.12.4` with pyright `1.1.369`
the following sample exhibits the same issue for me, in my case it is important to have the `AsyncGenerator` return type annotated.
```py
class Test(ABC):
    @abstractmethod
    async def test(self) -> AsyncGenerator[bool]:
        pass


class TestImpl(Test):
    async def test(self) -> AsyncGenerator[bool]:
        yield True
```

I've put this in the pyright playground as well: https://pyright-play.net/?code=GYJw9gtgBAhgRgYygSwgBzCALlAggIQGEAaWOAZyxBgSwgFMsALMAEwChRIosBPNZADsA5inSYcucr0EIA4vUH1qWTO3YIANjHLkoAFXqUAFAUIBKAFzsotqAAF4larQbM2NuzplJW9YDxGWMbk9JrA5lAAtAB8eNKyCkoqmADacGBgmgC61nb5UGg65Ooa2roGQQCS6JrGhpRWnrbeslB%2BAVhBIWER0XFSPknKMKog6Zk5eQW2vMhhrAYgAK707EA

**Reply by erictraut** [reply]
In Python, the presence of a `yield` statement changes the type of a function or method. Type checkers like pyright and mypy reflect this in their type evaluation logic.

There has been some [recent discussion](https://discuss.python.org/t/overloads-of-async-generators-inconsistent-coroutine-wrapping/56665) about clarifying and/or changing this in the typing spec, but no consensus has been reached. Unless/until the spec is modified, I don't plan to make any changes in pyright because the current behavior has been in place for a long time, and it matches mypy's behavior.

You can accommodate the current behavior by adding a `yield` statement to the abstract method implementation.

```python
class Test(ABC):
    @abstractmethod
    async def test(self) -> AsyncGenerator[bool]:
        yield True
```

**Reply by kavinyao** [reply]
@erictraut Does it mean it's not possible to avoid type error at the moment if the return type include a typevar (like [this](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwM4E8B2BjYEBQOBmUEIcMKYwaA5nKJLHAIKqYDiApmu1AjNADRwOXKMAyCAKuXYA1BFDw4JcALxwpYWfIAUAIgm6AlIowAbZEjgAhZO23DuYgNoSAuoYBcOOD8QsMcAAm7PhwVJzcvHZI7Kb4hnAAtAB8TP4OPHxQLoIAchBcrl6%2BJXA8wDFw%2BTAAkuCm7CCcMOyBAKJQRAp4ZhZwACKOAG6t2jYxTpQw7sW%2ByOgBwaHhIlHaMXEJKWkLGbzQk2gweQXsRd6lPvjQtLRoZQjUdgCMAAyeF5clKMCxgbSKeaYIIhOAgBCUbQfEpAgLXKC3SgDYajQwAOhWkRaUNmlzAoiO2mAxjwwFCAH1yWgEE1Kao1LpKeDKJTdLi-AtsGioABXNDaZn8wzGIA))?

**Reply by erictraut** [reply]
@kavinyao, if you're able to modify `Base.generate`, you can [add a `yield` statement](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwM4E8B2BjYEBQOBmUEIcMKYwaA5nKJLHAIKqYDiApmu1AjNADRwOXKMAyCAKuXYA1BFDw4JcALxwpYWfIAUAIgm6AlIowAbZEjgAhZO23DuYgNoSAuoYBcOOD8QsMcAAm7PhwVJzcvHZI7Kb4hnAAtAB8TP4OPHxQLoIAchBcrl6%2BJXA8wDFw%2BTAAkuCm7CCcMOyBAKJQRAqlvijAsYEm5kiWACKOAG6t2jYxTpQw7sW%2ByOgBwaHhIlHaMXEJKWlrGbzQ82gweQXsRd49cPjQtLRoZQjUdgCMAAyed-e9fqmQK0RSrTBBEJwEAISjaP4lcEBR5QZ6UODjURTQLwgB0W0iLXhyx6YFEF20wGMeGAoQA%2BnS0AgmgzVGpdAyYZQGboSX41thcVAAK5obRcsWGYxAA) after the `raise` statement. That will turn it into a coroutine, and `Derived.generate` will be compatible with it.

**Reply by kavinyao** [reply]
I thought I must yield a value but looks like it's not the case. Thanks!

**Reply by DimitriPapadopoulos** [reply]
In my case (https://github.com/zarr-developers/zarr-python/pull/2381) I have to silence MyPy as it complains about the value-less `yield`:
```python
class Test(ABC):
    @abstractmethod
    async def test(self) -> AsyncGenerator[bool]:
        yield  # type: ignore[misc]
```
